ENH: tolerance argument for limiting pad, backfill and nearest neighbor reindexing

shoyer opened the pull request:
xref #9817

This does not entirely solve the floating point precision issues, but gets us part of the way there -- we can explicitly lookup data with a fixed tolerance for nearest neighbor matches.

It is also is useful in its own right, mostly as a simple sanity check to verify that labels are not entirely misaligned.

Example usage:

```
In [2]: df = pd.DataFrame({'x': range(5)})

In [3]: df.reindex([0.1, 1.9, 3.5], method='nearest', max_distance=0.2)
Out[3]:
      x
0.1   0
1.9   2
3.5 NaN
```

shoyer committed:
ENH: add tolerance to get_indexer, get_loc and reindex

jreback commented on pandas/core/index.py:
```diff
@@ -1649,15 +1654,20 @@ def get_indexer(self, target, method=None, limit=None):
+        max_distance : optional
+            Maximum distance between target and index value for inexact
+            matches.
+
```
I think you should raise somewhere if ``max_distance`` is not None and you have an incompatible ``method`` (e.g. right now only 'nearest')

jreback commented:
``max_distance`` -> ``precision``? slightly more inline with other routines of this type (e.g. ``allclose``)

shoyer commented:
Hmm. Getting the name / API design right here is a little tricky.

My thought was that `max_distance` makes the most sense in terms of selecting nearest values.

On the other hand, `precision`/`tol` (or `atol`/`abs_tol`?) makes more sense in the context of imprecise matches / floating point issues. When providing a tolerance or precision, it seems implicit that you're doing nearest neighbor lookups, i.e., `precision=0.1` would imply `method='nearest'`.

For alignment purposes, we probably want an API that looks similar to [`math.isclose`](https://www.python.org/dev/peps/pep-0485/), which has the signature `isclose(a, b, rel_tol=1e-9, abs_tol=0.0)` and returns the result of `abs(a-b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)`.

shoyer commented:
Some further thoughts on API design (after some offline discussions):

- It's difficult to conceive of a true need for relative tolerance, because array labels are generally not multi-scale. A fixed absolute tolerance seems like the use case here.
- I'm leaning toward the argument name `tolerance`, which makes sense in the context of getting nearest values (at least somewhat) and also for alignment purposes -- we can later add a `tolerance` parameter to `align` and index set operations.

shoyer commented:
Here's an unsettled API design question: What should happen when you do `df.reindex(..., tol=1e-3)`?

In particular: should this imply `method='nearest'` if `method` is not provided? This is probably the behavior we'll want for align (e.g., `df1.align(df2, tol=1e-3)`) -- I don't think there's a usecase for one-sided tolerance in alignment.

The alternative (as @jreback suggests) is to just raise an error....

TomAugspurger commented:
> - A fixed absolute tolerance seems like the use case here.
> - I'm leaning toward the argument name tol/tolerance,

Agreed with both of these.
My only experience here is when I was working with lat/lngs which got misaligned from the regular grid I was expecting. A `tol` argument would have been perfect for me there.

shoyer commented:
Added some thoughts on possible API for setting a default tolerance on an index in #9817.

@TomAugspurger indeed, that's pretty much the exact case I'm thinking of here :).

wrobstory commented:
Feedback was solicited, so one thought: I generally prefer leaning towards the more-verbose keyword/argument names, unless it is a very, very long word. `tolerance` will be more easily deciphered when reading through someone else's code than `tol`.

jorisvandenbossche commented:
+1 on `tolerance`. If we are going to use it in other functions as well, this seems a better fit in general (although I find ``max_distance`` is a bit more explicit when it is only about 'nearest' reindexing)

Would this have a default value for FloatIndices? As I think they are only useful when that would be the case (otherwise all operations would become way too much verbose)

shoyer commented:
@jorisvandenbossche Yeah, I agree that some sort of default tolerance would be nice. The problem is that while it's true that many floating point index operations are only useful with a defined tolerance, how to automatically infer that tolerance is not obvious -- any solution seems likely to involve enough magic that it will break in unexpected ways.

jreback commented:
I think we could provide a tolerance as a keyword on the constructor and have a simple way of inferring (if needed) eg just do a couple of sample and diff and see

shoyer commented:
There is a subtle difference between `tolerance` and `max_distance` is if we generalize nearest neighbor indexing support to MultiIndex (#9365): the former implies an independent precision for each dimension, whereas the later implies a combined maximum distance. This distinction suggests to me that while `tolerance` would make more sense as an innate index property (e.g., as a keyword on the constructor), `max_distance` might make more sense as the keyword argument for `get_indexer`/`reindex`.

shoyer commented:
I have now renamed the parameter to `tolerance`, per the consensus. I have also made `method='nearest'` the default if a tolerance is supplied. We could also raise an error, but IMO `df.reindex(targets, tolerance=1e-3)` is not ambiguous, and we'll want to make the "nearest" aspect implicit for alignment operations.

jreback commented:
I don't think tolerance can be allowed for some inferred dtypes iow

strings and cases where tolerance is not a relevant unit , eg DatetimeInde/TimedeltaIndex/PeriodIndex must be a Timedelta or Offset

shoyer commented:
I think I have tests for other drupes -- I'll double check. 

On further consideration, I think I need to remove the default method='nearest' when a tolerance is provided. If I provide tolerance=2, I would expect a reindex/get_loc to match everything within that tolerance, not just the nearest value.

shoyer commented:
You now get errors if you try to use `limit` or `tolerance` with the default reindexing method.

With the current approach, you get a `ValueError` when you try to use `get_indexer` with an invalid tolerance, but `KeyError` on `DatetimeIndex.get_loc`. The later is because `get_loc` currently uses a rats nest of exception handling to allow for various types of indexers. Adding the proper error for an invalid tolerance there is probably a good idea but I'm not sure it's worth the trouble....

jreback commented on pandas/core/index.py:
```diff
@@ -1753,6 +1775,14 @@ def _get_nearest_indexer(self, target, limit):
+                target, indexer, tolerance)
+        return indexer
+
+    def _limit_indexer_tolerance(self, target, indexer, tolerance):
```
very confusing to have limit in the name.

shoyer replied:
renamed to `_filter_indexer_tolerance`

jreback commented on pandas/core/index.py:
```diff
@@ -1753,6 +1775,14 @@ def _get_nearest_indexer(self, target, limit):
+
+    def _limit_indexer_tolerance(self, target, indexer, tolerance):
+        distance = abs(self.values[indexer] - target)
+        indexer = np.where(distance <= tolerance, indexer, -1)
```
if the distance and the tolerance are invalid units, e.g. an integer on a ``DatetimeIndex`` this will give a very odd error message. Further this is not tested AFAICT.

shoyer replied:
Adding a bunch of error handling.

Unfortunately, we do allow integers currently as valid arguments to `Timedelta` (they are treated as nanoseconds). So I think we need to allow them as valid tolerance arguments, too...

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2755,10 +2789,27 @@ def test_get_loc(self):
+                                     tolerance=np.timedelta64(1, 'D')), 1)
+        self.assertEqual(idx.get_loc('2000-01-01T12', method='nearest',
+                                     tolerance=timedelta(1)), 1)
+        # TODO: fix this (it currently raises KeyError)
```
this should raise a nice ``ValueError``

jreback commented on pandas/tests/test_index.py:
```diff
@@ -2755,10 +2789,27 @@ def test_get_loc(self):
+        # with tm.assertRaises(ValueError):
+        #     idx.get_loc('2000-01-01T12', method='nearest', tolerance='foo')
+        with tm.assertRaises(KeyError):
+            idx.get_loc('2000-01-01T03', method='nearest',
```
the problem is ``_limit_indexer_tolerance`` is wrong

shoyer replied:
nope, this is a legit `KeyError` (2 hours is not enough tolerance for 3 hours offset)

shoyer commented:
OK, OK -- I added a bunch more error handling for `tolerance`.

shoyer commented:
@jreback any further comments?

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -80,6 +80,15 @@ Other enhancements
+- Reindex now has a ``tolerance`` argument that allows for finer control of :ref:`basics.reindex_fill`:
+
+  .. ipython:: python
+
```
would add a small section in ``missing.rst`` (and link to from here), that shows off ``tolerance/nearest``

shoyer replied:
wouldn't it belong here instead?
http://pandas.pydata.org/pandas-docs/stable/basics.html#basics-reindexing

shoyer replied:
actually I meant here: http://pandas.pydata.org/pandas-docs/stable/basics.html#filling-while-reindexing

jreback replied:
yes that's a much better location. points to the fact that there is a section in missing.rst that sort of deals with reindexing with filling, so maybe open an issue to clarify this (e.g. put pointers back-forth or something). I think the section is missing.rst is prob older.

jreback commented on pandas/core/generic.py:
```diff
@@ -1540,7 +1540,8 @@ def select(self, crit, axis=0):
-    def reindex_like(self, other, method=None, copy=True, limit=None):
+    def reindex_like(self, other, method=None, copy=True, limit=None,
+                     tolerance=None):
         """ return an object with matching indicies to myself
```
add to the doc-string

jreback commented on pandas/core/generic.py:
```diff
@@ -1725,7 +1727,12 @@ def sort_index(self, axis=0, ascending=True):
         limit : int, default None
-            Maximum size gap to forward or backward fill
+            Maximum number of consecutive elements to forward or backward fill
+        tolerance : optional
```
this is in the reindex doc-string yes?

shoyer replied:
yes

jreback commented on pandas/tests/test_index.py:
```diff
@@ -1104,6 +1104,15 @@ def test_get_indexer(self):
         assert_almost_equal(r2, e1[::-1])
 
+    def test_get_indexer_invalid(self):
+        idx = Index(np.arange(10))
```
add the issue number as a comment

jreback commented on pandas/core/index.py:
```diff
@@ -1753,6 +1785,14 @@ def _get_nearest_indexer(self, target, limit):
+        return indexer
+
+    def _filter_indexer_tolerance(self, target, indexer, tolerance):
+        distance = abs(self.values[indexer] - target)
```
I think you need to call ``_convert_tolerance`` here (you do this in ``tseries/base/get_loc`` ) so not sure which is the correct convention

shoyer replied:
Nope -- this is only called from inside `get_indexer`, which already calls `_convert_tolerance` at the top. The extra call to `_convert_tolerance` in `DatetimeIndex.get_loc` is to ensure that an appropriate error gets raised for a bad tolerance argument that cannot be coerced to a timedelta.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -80,6 +80,15 @@ Other enhancements
+     df = pd.DataFrame({'x': range(5)})
+     df.reindex([0.1, 1.9, 3.5], method='nearest', tolerance=0.2)
+
+  ``tolerance`` is also exposed by the lower level ``Index.get_indexer`` and ``Index.get_loc`` methods.
```
also would add an example maybe with a datetimelike (and timedelta tolerance)

jreback commented on pandas/tseries/period.py:
```diff
@@ -664,6 +664,9 @@ def _get_string_slice(self, key):
                      self.searchsorted(t2.ordinal, side='right'))
 
+    def _convert_tolerance(self, tolerance):
+        raise NotImplementedError('tolerance not supported for PeriodIndex')
```
? why not I don't we support diff on PeriodIndex?

shoyer commented on pandas/core/index.py:
```diff
@@ -1677,33 +1690,52 @@ def get_indexer(self, target, method=None, limit=None):
+                                     tolerance=tolerance)
+
+        if tolerance is not None:
+            tolerance = self._convert_tolerance(tolerance)
```
@jreback this is where I do the tolerance coercion... trying to remember why I moved it up here, I think this was to ensure it will be coerced properly before possibly converting the index to object dtype (on the lines below). I'll add a comment.

jreback replied:
yeh I saw that. But I don't think this gets called if only ``get_indexer`` is called.

jreback commented:
looks pretty good.

not sure if this matters, but something has to be: ``value +- test_value <= tolerance`` right? I would just make equal clear, that a value that is equal to the tolerance is INCLUDED.

shoyer commented:
@jreback OK, I think I responded to all your suggestions:

1. This now works on `PeriodIndex` (was easier than I thought)
2. Added a section on `limit` and `tolerance` to the "Filling while reindexing" docs in basics.rst
3. Added an example of using tolerance with a datetimeindex to what's new
4. Updated docstrings to clarify the formula used for `tolerance`.

This still needs a rebase/squash, but I thought I would get your review before squashing.

jreback commented on doc/source/basics.rst:
```diff
@@ -1100,6 +1100,25 @@ Note that the same result could have been achieved using
+The ``limit`` and ``tolerance`` arguments provide even more control over
+filling while reindexing. Limit specifies the maximum count of consecutive
+matches:
+
```
make a separate sub section

shoyer replied:
done

jreback commented:
lgtm
minor comment
rebase / squash snd merge

shoyer commented:
OK, did the rebase / squash.

If Travis gives the OK, I'll merge EOD tomorrow.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -1560,7 +1561,10 @@ def reindex_like(self, other, method=None, copy=True, limit=None):
+            Maximum number of consecutive labels to fill for inexact matches.
+        tolerance : optional
+            Maximum distance between labels of the other object and this
+            object for inexact matches.
```
Can you add a versionadded here?

shoyer replied:
done

shoyer commented:
Travis passed prior to the last docstring fixup, so I decided it was safe to merge.

Thanks everyone for your help on this one!

jreback commented:
nice :)
